draft-ietf-tls-dtls13.md

```diff
@@ -347,6 +392,52 @@ allowing the sequence number to wrap.
 Implementations MUST NOT allow the epoch to wrap, but instead MUST establish
 a new association, terminating the old association.
 
+### Determining the Header Format
+
+Implementations can distinguish the two header formats by examining
+the first byte, which in the DTLSCiphertext header represents the
+content type. The only valid values for the content type are
+alert(21), handshake(22), application_data(23), and ack(25),
+with application_data actually
+representing protected data with the true content type being
+encrypted. If any of these values is present, then the record MUST be
+handled as DTLSCiphertext.
+
+If the first byte is any other other value, then receivers
+MUST check to see if the leading bits of the first byte are
+001. If so, they MUST process the record as DTLSShortCiphertext.
+Otherwise, the record MUST be rejected as if it had failed
+deprotection.
+
+### Reconstructing the Sequence Number and Epoch {#reconstructing}
+
+When receiving a DTLSShortCiphertext message, the recipient does not
+have a full epoch or sequence number value and so there is some
+opportunity for ambiguity.  Because the full epoch and sequence number
+are used to compute the per-record nonce, failure to reconstruct these
+values leads to failure to deprotect the record, and so implementations
+MAY use a mechanism of their choice to determine the full values.
+This section provides an algorithm which is comparatively simple
+and which implementations are RECOMMENDED to follow.
+
+If the epoch bit has the same parity as the current epoch, then
+implementations SHOULD reconstruct the sequence number by computing
+the full sequence number which is numerically closest to the most
+recently successfully deprotected record.
+
+If the epoch bit has a different parity from the current epoch, then
+the record is either from a previous epoch or from a future
+epoch. Implementations SHOULD use the epoch value which would produce
+a sequence number which is numerically closest, as determined by the
+algorithm in the paragraph above.
```

Comment:
Are you suggesting that the next and previous epoch are checked?  Or maybe all potentially active epochs?  If so, you need to mention that the next sequence number is from the selected epoch and not the current one.

I think that we can avoid some of the potential for variation with a different algorithm (noting of course that you permit variations).  If we restrict KeyUpdate to only one outstanding change, then you only ever have to select between two epochs.  If you require acknowledgment of KeyUpdate before switching epochs, then you know to expect a change.

Then, without KeyUpdate, you have just one epoch and can discard records that don't match parity.

With KeyUpdate (which you should put in a new section), you have two epochs active from the moment that you send an acknowledgment for a KeyUpdate.  You can select between them using the bit.

Reply:
I agree with these comments. Note the OPEN ISSUE markers below. I was hoping to just get this in.